### PR TITLE
feat: Make torch-cluster dependency optional

### PR DESCRIPTION
**Pull request by HennerM**
We are using torch_geometric and one of it's libraries torch-cluster for the graph operations in Supernode pooling. 

torch-cluster comes with PyTorch extensions providing custom C++ and CUDA implementations. These should be able to give better performance than what is possible with native Python/Pytorch. However, installing these dependencies is not trivial because we either need to install prebuilt binaries that depend on the specific version of Python,PyTorch and CUDA installed or we install from source, which requires a working C++ and optionally CUDA compiler and torch headers + shared libraries (which are included in the Pytorch pip package). This is the reason we need to install pytorch first and run pip install with --no-build-isolation.

For ease of use it makes sense to make torch-cluster and thus torch-geometric optional, since we can also use naive PyTorch operations to implement the used operators, even if we have to take a slight performance hit.

For CUDA performance I implemented Triton kernels that are based on torch-cluster, spawning one thread per `y` point. These have the advantage that they are JIT compiled triton comes with newer PyTorch versions anyway.

I did some naive benchmarking on the implementations, comparing to the PyG torch-cluster ones:

```
----------------------------------------------------------------------------------------- benchmark 'knn': 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                            Min                 Max                Mean            StdDev              Median                IQR            Outliers      OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_performance_knn[triton-cuda]        16.9356 (1.0)       17.0483 (1.0)       16.9755 (1.0)      0.0426 (1.05)      16.9659 (1.0)       0.0321 (1.0)           1;1  58.9084 (1.0)           5           1
test_performance_knn[pyg-cuda]           19.7955 (1.17)      20.0943 (1.18)      19.9128 (1.17)     0.0758 (1.87)      19.9577 (1.18)      0.1102 (3.44)         13;0  50.2189 (0.85)         43           1
test_performance_knn[pyg-cpu]            69.1496 (4.08)      69.5319 (4.08)      69.2973 (4.08)     0.1057 (2.60)      69.2706 (4.08)      0.1121 (3.49)          4;1  14.4306 (0.24)         15           1
test_performance_knn[fallback-cpu]      197.3092 (11.65)    213.7731 (12.54)    206.7972 (12.18)    7.3766 (181.69)   207.9078 (12.25)    13.4083 (417.83)        1;0   4.8357 (0.08)          5           1
test_performance_knn[fallback-cuda]     202.7169 (11.97)    202.8215 (11.90)    202.7569 (11.94)    0.0406 (1.0)      202.7464 (11.95)     0.0524 (1.63)          1;0   4.9320 (0.08)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------- benchmark 'radius': 5 tests -----------------------------------------------------------------------------------------
Name (time in ms)                               Min                 Max                Mean            StdDev              Median               IQR            Outliers       OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_performance_radius[triton-cuda]         1.9012 (1.0)        2.0247 (1.0)        1.9473 (1.0)      0.0486 (2.74)       1.9326 (1.0)      0.0638 (7.07)          1;0  513.5246 (1.0)           5           1
test_performance_radius[pyg-cuda]            3.8028 (2.00)       3.8963 (1.92)       3.8251 (1.96)     0.0177 (1.0)        3.8240 (1.98)     0.0090 (1.0)           2;2  261.4307 (0.51)         21           1
test_performance_radius[pyg-cpu]           119.9390 (63.09)    121.0944 (59.81)    120.3918 (61.82)    0.4424 (24.94)    120.3057 (62.25)    0.7428 (82.29)         3;0    8.3062 (0.02)          8           1
test_performance_radius[fallback-cuda]     201.7581 (106.12)   202.0828 (99.81)    201.8424 (103.65)   0.1360 (7.67)     201.7956 (104.42)   0.1149 (12.73)         1;1    4.9544 (0.01)          5           1
test_performance_radius[fallback-cpu]      238.3177 (125.35)   246.2958 (121.64)   242.4150 (124.49)   2.9777 (167.86)   242.7741 (125.62)   3.9730 (440.18)        2;0    4.1252 (0.01)          5           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
================================================================================================= 10 passed, 8 skipped, 1 warning in 25.99s ====================================================================

```

Obviously we would have to benchmark over more dimensions since this is just covering one case now, but performance is not the main motivation here anyway.

### CODE REVIEW

**Review by Ndles — DISMISSED** [review]

File: src/noether/modeling/functional/geometric.py [hunk 1]
```diff
@@ -1,10 +1,377 @@
...
+    # Validate batch tensors
+    if batch_x is not None and batch_y is not None:
+        batch_x = batch_x.contiguous().int()
+        batch_y = batch_y.contiguous().int()
+        assert batch_x.shape[0] == n_x
+        assert batch_y.shape[0] == n_y
+    else:
+        # Create dummy tensors for kernel
+        batch_x = torch.zeros(1, dtype=torch.int32, device=x.device)
+        batch_y = torch.zeros(1, dtype=torch.int32, device=y.device)
```
Comment (Ndles) [inline src/noether/modeling/functional/geometric.py]: Maybe specifying a proper tensor shape for the kernel makes sense, WDYT?
```suggestion
has_batch = batch_x is not None and batch_y is not None
if has_batch:
    batch_x = batch_x.contiguous().int()
    batch_y = batch_y.contiguous().int()
else:
    batch_x = torch.zeros((n_x,), dtype=torch.int32, device=x.device)
    batch_y = torch.zeros((n_y,), dtype=torch.int32, device=y.device)
```
